File: Design-Patterns-Using-JavaScript/Singleton/LimitedInstances.js
class LimitedInstances {
    static instances = [];
    static limit = 4;

    constructor() {}

    static getInstance() {
        if (LimitedInstances.instances.length < LimitedInstances.limit) {
            const instance = new LimitedInstances();
            LimitedInstances.instances.push(instance);
            return instance;
        } else {
            throw new Error("Instance Limit reached");
        }
    }
}

// Client code
try {
    LimitedInstances.getInstance();
    LimitedInstances.getInstance();
    LimitedInstances.getInstance();
    LimitedInstances.getInstance();
    LimitedInstances.getInstance();
    LimitedInstances.getInstance();
    LimitedInstances.getInstance();
    LimitedInstances.getInstance();
    LimitedInstances.getInstance();
    LimitedInstances.getInstance();
} catch (error) {
    console.log(error.message);
}

File: Design-Patterns-Using-Ruby/Singleton/LimitedInstances.rb
class LimitedInstances
  @@instances = []
  LIMIT = 4

  private_class_method :new

  def self.get_instance
    if @@instances.length < LIMIT
      instance = new
      @@instances.push(instance)
      puts 'Instance created.'
      instance
    else
      raise 'Instance Limit reached.'
    end
  end
end

# Client code
begin
  LimitedInstances.get_instance
  LimitedInstances.get_instance
  LimitedInstances.get_instance
  LimitedInstances.get_instance
  LimitedInstances.get_instance
  LimitedInstances.get_instance
  LimitedInstances.get_instance
  LimitedInstances.get_instance
  LimitedInstances.get_instance
  LimitedInstances.get_instance
rescue StandardError => e
  puts e.message
end
